packages/async-flow/src/replay-membrane.js

```diff
@@ -0,0 +1,393 @@
+    if (status === 'settled') {
+      return;
+    }
+    const guestFulfillment = hostToGuest(hostFulfillment);
+    status.resolve(guestFulfillment);
+    guestPromiseMap.set(guestPromise, 'settled');
+  };
+
+  /**
+   * When replaying, this comes from interpreting the log.
+   * Otherwise, it is triggered by a watcher watching hostVow,
+   * that must also log it.
+   *
+   * @param {HostVow} hostVow
+   * @param {Host} hostReason
+   */
+  const doReject = (hostVow, hostReason) => {
+    const guestPromise = hostToGuest(hostVow);
+    const status = guestPromiseMap.get(guestPromise);
+    if (status === 'settled') {
+      return;
+    }
+    const guestReason = hostToGuest(hostReason);
+    status.reject(guestReason);
+    guestPromiseMap.delete(guestPromise);
+  };
+
+  /**
+   * When replaying, after the guest thinks it has called a host method,
+   * triggering `checkCall`, that host method emulator consumes one of
+   * these entries from the log to return what it is supposed to.
+   * It returns an Outcome describing either a throw or return, because we
+   * reserve the actual throw channels for replay errors and internal
+   * errors.
+   *
+   * @param {number} callIndex
+   * @param {Host} hostResult
+   * @returns {Outcome}
+   */
+  const doReturn = (callIndex, hostResult) => {
+    unnestInterpreter(callIndex);
+    const guestResult = hostToGuest(hostResult);
+    return harden({
+      kind: 'return',
+      result: guestResult,
+    });
+  };
+
+  /**
+   * When replaying, after the guest thinks it has called a host method,
+   * triggering `checkCall`, that host method emulator consumes one of
+   * these entries from the log to return what it is supposed to.
+   * It returns an Outcome describing either a throw or return, because we
+   * reserve the actual throw channels for replay errors and internal
+   * errors.
+   *
+   * @param {number} callIndex
+   * @param {Host} hostProblem
+   * @returns {Outcome}
+   */
+  const doThrow = (callIndex, hostProblem) => {
+    unnestInterpreter(callIndex);
+    const guestProblem = hostToGuest(hostProblem);
+    return harden({
+      kind: 'throw',
+      problem: guestProblem,
+    });
+  };
+
+  // ///////////// Guest to Host or consume log ////////////////////////////////
+
+  const performCall = (hostTarget, optVerb, hostArgs, callIndex) => {
+    let hostResult;
+    try {
+      hostResult = optVerb
+        ? hostTarget[optVerb](...hostArgs)
+        : hostTarget(...hostArgs);
+      // Try converting here just to route the error correctly
+      hostToGuest(hostResult, `converting ${optVerb || 'host'} result`);
+    } catch (hostProblem) {
+      return logDo(nestDispatch, harden(['doThrow', callIndex, hostProblem]));
+    }
+    return logDo(nestDispatch, harden(['doReturn', callIndex, hostResult]));
+  };
+
+  const guestCallsHost = (guestTarget, optVerb, guestArgs, callIndex) => {
+    if (!bijection.hasGuest(guestTarget)) {
```

Comment:
It might be redundant, but maybe this should check `!stopped` as well ?

Reply:
Done